### PR TITLE
Apply [[noreturn]] to Kokkos::abort when applicable

### PR DESCRIPTION
**Pull request by AndrewGaspar**
I haven't run any tests beyond just building. Pushing to run CI.

I removed the `__APPLE__` check for `cuda_abort` since I don't think macOS supports CUDA in recent versions anyway. I can put it back if this configuration is still supported, though.

### CODE REVIEW

**Comment by dalg24** [comment]
OK to test

**Comment by masterleinad** [comment]
> I removed the `__APPLE__` check for `cuda_abort` since I don't think macOS supports CUDA in recent versions anyway. I can put it back if this configuration is still supported, though.

I would say it does at least for CUDA versions we support: https://www.nvidia.com/en-us/drivers/cuda/418_163/macosx-cuda-418_163-driver/ and `assert()` is not supported on `MacOS` at least in 9.0 https://docs.nvidia.com/cuda/archive/9.0/cuda-c-programming-guide/index.html#assertion.

**Comment by AndrewGaspar** [comment]
OK, if people are still using CUDA on Mac, then I'll put it back.

File: core/src/impl/Kokkos_Error.hpp [hunk 1]
```diff
@@ -164,9 +164,23 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
 //----------------------------------------------------------------------------
 //----------------------------------------------------------------------------
 
+#if defined(KOKKOS_ENABLE_CUDA) && defined(__CUDA_ARCH__)
+// CUDA aborts
+#define KOKKOS_ABORT_NORETURN [[noreturn]]
```
Comment (dhollman) [inline core/src/impl/Kokkos_Error.hpp:169]:
```suggestion
#define KOKKOS_IMPL_ABORT_NORETURN [[noreturn]]
```

File: core/src/impl/Kokkos_Error.hpp [hunk 2]
```diff
@@ -164,9 +164,23 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
...
+#elif defined(KOKKOS_ENABLE_HIP) && defined(__HIP_DEVICE_COMPILE__)
+// HIP does not abort
+#define KOKKOS_ABORT_NORETURN
```
Comment (dhollman) [inline core/src/impl/Kokkos_Error.hpp:172]:
```suggestion
#define KOKKOS_IMPL_ABORT_NORETURN
```

File: core/src/impl/Kokkos_Error.hpp [hunk 3]
```diff
@@ -164,9 +164,23 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
...
+#elif !defined(KOKKOS_ENABLE_OPENMPTARGET) && !defined(__HCC_ACCELERATOR__)
+// Host aborts
+#define KOKKOS_ABORT_NORETURN [[noreturn]]
```
Comment (dhollman) [inline core/src/impl/Kokkos_Error.hpp:175]:
```suggestion
#define KOKKOS_IMPL_ABORT_NORETURN [[noreturn]]
```

File: core/src/impl/Kokkos_Error.hpp [hunk 4]
```diff
@@ -164,9 +164,23 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
...
+#else
+// Everything else does not abort
+#define KOKKOS_ABORT_NORETURN
```
Comment (dhollman) [inline core/src/impl/Kokkos_Error.hpp:178]:
```suggestion
#define KOKKOS_IMPL_ABORT_NORETURN
```

File: core/src/impl/Kokkos_Error.hpp [hunk 5]
```diff
@@ -164,9 +164,23 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
...
+#endif
+
 namespace Kokkos {
-KOKKOS_INLINE_FUNCTION
-void abort(const char *const message) {
+KOKKOS_ABORT_NORETURN KOKKOS_INLINE_FUNCTION void abort(
```
Comment (dhollman) [inline core/src/impl/Kokkos_Error.hpp:182]:
```suggestion
KOKKOS_IMPL_ABORT_NORETURN KOKKOS_INLINE_FUNCTION void abort(
```

File: core/src/Cuda/Kokkos_Cuda_abort.hpp [hunk 6]
```diff
@@ -56,21 +56,20 @@ extern "C" {
...
+                                          unsigned int line,
+                                          const void *function,
+                                          size_t charsize);
 }
 
 namespace Kokkos {
 namespace Impl {
 
-__device__ inline void cuda_abort(const char *const message) {
-#ifndef __APPLE__
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_abort.hpp:68]: Why did you remove that check?

File: core/src/Cuda/Kokkos_Cuda_abort.hpp [hunk 7]
```diff
@@ -56,21 +56,20 @@ extern "C" {
 /*  Cuda runtime function, declared in <crt/device_runtime.h>
  *  Requires capability 2.x or better.
  */
-extern __device__ void __assertfail(const void *message, const void *file,
-                                    unsigned int line, const void *function,
-                                    size_t charsize);
+[[noreturn]] __device__ void __assertfail(const void *message, const void *file,
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: I still see
```
extern __host__ __device__ void __assertfail(const char * __assertion,
                                             const char *__file,
                                             unsigned int __line,
                                             const char *__function,
                                             size_t charsize);
```
in `crt/common_functions.h`. Adding `[[noreturn]]` here sounds dangerous.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 6 above)
Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:68]: I'll fix this, sorry, I was just trying to make sure all the configurations in CI passed.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 7 above)
Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: I think __assertfail is necessarily noreturn, right? Or is that not true? [[noreturn]] shouldn't change the ABI of the routine...

Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: Oh, is the concern that we're changing the definition for the routine? Hm... Maybe I'll just add an infinite loop after the __assertfail call instead...

Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: https://en.cppreference.com/w/cpp/language/attributes/noreturn says:
```
If a function is declared with [[noreturn]] in one translation unit, and the same function is declared without [[noreturn]] in another translation unit, the program is ill-formed; no diagnostic required.
```
That's my only concern.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:68]: Thanks!

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 7 above)
Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: Yeah, I think you're right.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 7 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: I assume the infinite loop suggestion was a joke right?

Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:59]: Well, no, but maybe you can help illuminate any potential issue. `__assertfail` is semantically `[[noreturn]]`, but it's not marked as such. You need to prove to the compiler that a function actually is no return, so putting an infinite loop after the `__assertfail` would do that. However, this loop would never be entered. Could it still cause issues?

File: core/src/Cuda/Kokkos_Cuda_abort.hpp (hunk 6 above)
Comment (AndrewGaspar) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:68]: Fixed.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp [hunk 8]
```diff
@@ -64,14 +64,24 @@ extern __device__ void __assertfail(const void *message, const void *file,
...
   const char empty[] = "";
 
   __assertfail((const void *)message, (const void *)empty, (unsigned int)0,
                (const void *)empty, sizeof(char));
-#endif
+
+  // This loop is never executed. It's intended to suppress warnings that the
+  // function returns, even though it does not. This is necessary because
+  // __assertfail is not marked as [[noreturn]], even though it does not return.
+  while (true)
```
Comment (AndrewGaspar) [inline core/src/Cuda/Kokkos_Cuda_abort.hpp:77]: @crtrott Do you have a better idea of how to do this if it is to be done?

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_abort.hpp:77]: no its probably fine

**Comment by crtrott** [comment]
Retest This Please

File: core/src/impl/Kokkos_Error.hpp [hunk 9]
```diff
@@ -164,9 +164,30 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
...
+#if defined(__APPLE__)
+// cuda_abort does not abort when building for macOS.
+#define KOKKOS_IMPL_ABORT_NORETURN
+#else
+// cuda_abort aborts when building for other platforms than macOS
+#define KOKKOS_IMPL_ABORT_NORETURN [[noreturn]]
+#endif
+
+#elif defined(KOKKOS_ENABLE_HIP) && defined(__HIP_DEVICE_COMPILE__)
+// HIP does not abort
```
Comment (masterleinad) [inline core/src/impl/Kokkos_Error.hpp:178]: This changed recently, see https://github.com/kokkos/kokkos/blob/89e9d26d58f94547ba219a71ed5bde4d935715cd/core/src/HIP/Kokkos_HIP_Abort.hpp#L56-L63

Comment (crtrott) [reply core/src/impl/Kokkos_Error.hpp:178]: oh man. I was about to merge it :-(

Comment (masterleinad) [reply core/src/impl/Kokkos_Error.hpp:178]: I would be fine with addressing that in a follow-up pull request.

Comment (masterleinad) [reply core/src/impl/Kokkos_Error.hpp:178]: I am currently seeing
```
/tmp/kokkos/core/src/HIP/Kokkos_HIP_Abort.hpp:64:1: warning: function declared 'noreturn' should not return [-Winvalid-noreturn]
```
if trying to mark the function with `[[noreturn]]` so we might just postpone fixing it.
